10_rest/nickolee/cmd/puppy-server/main.go

```diff
@@ -0,0 +1,122 @@
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"strconv"
+
+	"github.com/anz-bank/go-course/10_rest/nickolee/pkg/puppy/store"
+
+	puppy "github.com/anz-bank/go-course/10_rest/nickolee/pkg/puppy"
+	rest "github.com/anz-bank/go-course/10_rest/nickolee/pkg/puppy/rest"
+	"github.com/go-chi/chi"
+	"github.com/go-chi/chi/middleware"
+	"gopkg.in/alecthomas/kingpin.v2"
+)
+
+var out io.Writer = os.Stdout
+
+var (
+	args = os.Args[1:]
+	// this loads your file once parseFlags() is run
+	// these are methods too prepare a variable. Basically telling kingpin how to read the flag from cli
+	fileName = kingpin.Flag("data", "This flag supplies a puppy JSON file name").
+			Short('d').Default("../../puppy-data/puppies.json").File()
+	port   = kingpin.Flag("port", "Port to listen on and serve").Short('p').Default("7777").Int()
+	storer = kingpin.Flag("store", "Map/Sync Store").Short('s').Default("map").Enum("map", "sync")
+)
+
+// Config contains arguments parsed from commandline flags
+type Config struct {
+	file   *os.File
+	port   int
+	storer string
+}
+
+func main() {
+	// parse flags/check if user even provided any
+	cliConfig, err := parseFlags(args)
+	if err != nil {
+		fmt.Println(err)
+	}
+
+	// read JSON from file + parse/unmarshal the json and we now have []puppy (Go objects)
+	puppies, err := readFileAndUnmarshalPuppies(cliConfig.file)
+	if err != nil {
+		panic(err)
+	}
+
+	// instantiate new PuppyHandlerAndStorer based on user specified value of storer
+	// create puppy.Storer
+	newStorer, _ := createStore(cliConfig.storer) // no need to check err as it defaults to "map"
+
+	// create PuppyHandlerAndStorer - the Rest API wrapper around puppy.Storer
+	phs := rest.NewPuppyHandlerAndStorer(newStorer)
+
+	// get the user specified port
+	addr := ":" + strconv.Itoa(cliConfig.port)
+
+	// create some puppies loaded in from json file and store in store (no pun intended lol)
+	for _, pup := range puppies {
+		pup := pup // to avoid scopelint error: Using a reference for the variable on range scope `pup`
+		createdPuppy, _ := phs.Storage.CreatePuppy(&pup)
+		fmt.Fprintf(out, "Puppy with ID %d has been created\n", createdPuppy)
+		retrievedPuppy, _ := phs.Storage.ReadPuppy(createdPuppy)
+		fmt.Fprintln(out, "Retrieved puppy:", retrievedPuppy)
+	}
+
+	// setting up REST API wrapper
+	r := chi.NewRouter()
+	r.Use(middleware.RequestID)
+	r.Use(middleware.Logger)
+
+	rest.SetupRoutes(r, *phs)
+
+	fmt.Printf("Starting server on port %s. Try:\n", addr)
+	fmt.Printf("  curl localhost%s/api/puppy -d ", addr)
```

Comment:
You don't normally see output like this when you start a server. Consider the following.
1. What if it's not running on localhost
2. What if you didn't provide a datafile (getting puppy 4 would fail)

Also, you are posting a puppy, so it needs `curl -X POST stuff`, though I wouldn't keep it.

Reply:
ok will delete this perhaps move to the readme